File: server/cost/advisor.js
// Cost-at-decision-time (root CLAUDE.md's Operational Awareness item 6) —
// distinct from report.js's HISTORICAL spend tracking: this feeds a real,
// measured cost signal back into models/router.js's own ranking, which
// already weighs `entry.tier.cost` in every one of its scoring branches
// (background work is already the cost-heaviest, at -2x). What was missing
// wasn't the weighting — it's that `tier.cost` was always a 1-5 name-regex
// GUESS (catalog.js's guessFromName()), never a measured fact.
//
// Deliberately returns a value in the EXACT SAME 0-4 domain catalog.js's
// tier.cost already uses, rather than inventing a new scale or a raw dollar
// adjustment — router.js's scoreFor() was already built and tuned around a
// 0-4 tier.cost value multiplied by small fixed weights (-0.5x to -2x
// depending on branch), all comfortably inside the ±20
// AVAILABILITY_SCORE_BONUS spread router.js's own header comment documents
// as the thing nothing else may ever exceed. A same-domain replacement
// value can mathematically never break that bound — it was already proven
// safe for the exact range this returns.
//
// Returns null (never a guess of its own) when there isn't yet enough real
// data to correct the catalog's guess — router.js falls back to the
// catalog's own tier.cost in that case, exactly as it always has.

import { getPrice } from './cost-store.js';

// $ per 1K tokens (a blended, rough in/out average — this is a coarse
// bucket assignment, not a precise cost estimate; report.js's own
// usageBreakdown() is where a precise, real dollar figure lives) mapped
// onto catalog.js's existing 0-4 cost tier scale.
const BUCKETS = [
  { maxPer1k: 0.0005, cost: 0 },
  { maxPer1k: 0.002, cost: 1 },
  { maxPer1k: 0.01, cost: 2 },
  { maxPer1k: 0.03, cost: 3 },
  { maxPer1k: Infinity, cost: 4 },
];

function bucketFor(usdPer1k) {
  for (const b of BUCKETS) {
    if (usdPer1k <= b.maxPer1k) return b.cost;
  }
  return 4;
}

/**
 * A real-price-derived 0-4 cost tier for one (provider, modelId), or null
 * if no real price is on record yet — see this file's header comment for
 * why null (not a fallback number) is the honest thing to return here.
 */
export function observedCostTier(provider, modelId) {
  if (!provider || !modelId) return null;
  const price = getPrice(provider, modelId, 'tokens');
  if (!price) return null;
  const priceIn = typeof price.priceIn === 'number' ? price.priceIn : 0;
  const priceOut = typeof price.priceOut === 'number' ? price.priceOut : 0;
  // Both zero is a genuine free/local model, not "no data" — bucket 0 is
  // correct for it, not null.
  const blendedPer1k = ((priceIn + priceOut) / 2) * 1000;
  return bucketFor(blendedPer1k);
}

File: backend/jarvis/cost/advisor.py
"""Cost at decision time: letting the router rank on a MEASURED price rather
than a guess.

The router already weighed `tier.cost` in every scoring branch. What it was
weighing was a 1-5 guess made by matching the model's NAME against a few
regexes. This substitutes a value derived from a real recorded price — in the
exact same 0-4 domain the guess already used, deliberately, so the substitution
is mathematically guaranteed to stay inside the score spread the availability
bonus was sized against. A new scale would have needed the whole ranking
re-tuned to stay safe.

Returns None when there is no real price yet, and the router keeps the catalog's
own guess. A fallback number invented here would be a third guess wearing the
authority of a measurement.
"""

from __future__ import annotations

import time
from typing import Any

from . import store

#: USD per 1,000 tokens, blended across input and output. A coarse bucket, not a
#: cost estimate — `report.py` is where a real figure lives.
BUCKETS: tuple[tuple[float, int], ...] = (
    (0.0005, 0), (0.002, 1), (0.01, 2), (0.03, 3),
)

#: Scoring runs per candidate per turn, so an uncached read would mean a SQLite
#: query per model per turn. Prices change on a daily refresh at most.
CACHE_TTL_S = 60.0
_cache: dict[tuple[str, str], tuple[float, int | None]] = {}


def _bucket_for(usd_per_1k: float) -> int:
    for ceiling, tier in BUCKETS:
        if usd_per_1k <= ceiling:
            return tier
    return 4


def observed_cost_tier(provider: str | None, model_id: str | None) -> int | None:
    if not provider or not model_id:
        return None
    key = (provider, model_id)
    cached = _cache.get(key)
    now = time.monotonic()
    if cached and now - cached[0] < CACHE_TTL_S:
        return cached[1]

    price = store.get_price(provider, model_id, "tokens")
    tier: int | None = None
    if price is not None:
        price_in = price["priceIn"] if isinstance(price["priceIn"], (int, float)) else 0.0
        price_out = price["priceOut"] if isinstance(price["priceOut"], (int, float)) else 0.0
        # Both zero is a genuine free or local model, not missing data: bucket 0
        # is the right answer for it, and None would be wrong.
        tier = _bucket_for(((price_in + price_out) / 2) * 1000)
    _cache[key] = (now, tier)
    return tier


def reset_cache() -> None:
    """After a price write, and in tests. Cheap: the cache is tiny."""
    _cache.clear()
